vhost-user-backend/src/handler.rs

```diff
@@ -617,6 +632,68 @@ where
 
         Ok(())
     }
+
+    #[cfg(not(feature = "xen"))]
+    fn postcopy_advice(&mut self) -> VhostUserResult<File> {
+        let mut uffd_builder = UffdBuilder::new();
+
+        let uffd = uffd_builder
+            .close_on_exec(true)
+            .non_blocking(true)
+            .user_mode_only(false)
+            .create()
+            .map_err(|e| {
+                VhostUserError::ReqHandlerError(io::Error::new(io::ErrorKind::Other, e))
+            })?;
+
+        // We need to duplicate the uffd fd because we need both
+        // to return File with fd and store fd inside uffd.
+        //
+        // SAFETY:
+        // We know that uffd is correctly created.
+        // This means fd inside uffd is also a valid fd.
+        // Duplicating a valid fd is safe.
+        let uffd_dup = unsafe { libc::dup(uffd.as_raw_fd()) };
+        if uffd_dup < 0 {
+            return Err(VhostUserError::ReqHandlerError(io::Error::last_os_error()));
+        }
+
+        // SAFETY:
+        // We know that uffd_dup is a valid fd.
+        let uffd_file = unsafe { File::from_raw_fd(uffd_dup) };
+
+        self.uffd = Some(uffd);
+
+        Ok(uffd_file)
+    }
+
+    #[cfg(not(feature = "xen"))]
+    fn postcopy_listen(&mut self) -> VhostUserResult<()> {
+        let Some(ref uffd) = self.uffd else {
+            return Err(VhostUserError::ReqHandlerError(io::Error::new(
+                io::ErrorKind::Other,
+                "No registered UFFD handler",
+            )));
+        };
+
+        for mapping in self.mappings.iter() {
+            uffd.register(
+                mapping.local_addr as *mut libc::c_void,
+                mapping.size as usize,
+            )
+            .map_err(|e| {
+                VhostUserError::ReqHandlerError(io::Error::new(io::ErrorKind::Other, e))
+            })?;
+        }
+
+        Ok(())
+    }
+
+    #[cfg(not(feature = "xen"))]
+    fn postcopy_end(&mut self) -> VhostUserResult<()> {
+        _ = self.uffd.take();
```

Comment:
If we are not using the return value, what about using `self.uffd = None` ?

Reply:
I think the `Option::take` is more idiomatic. Basically we assign whatever is in the `uffd` to `_` and it will be dropped at the end of the scope.

Reply:
IMHO it's clearer by assigning `None` that we're basically dropping whatever value was there before. It might make sense to use `take()` if we then check that there really was something there (should we do that and return an error?).

Reply:
Ok, switched to `self.uffd = None`. I don't think there should be an error if uffd does not exist. Qemu does not error out in this case as well.